client/src/main/java/io/pravega/client/stream/impl/SegmentReaderImpl.java

```diff
@@ -0,0 +1,97 @@
+import io.pravega.client.segment.impl.EndOfSegmentException;
+import io.pravega.client.segment.impl.EventSegmentReader;
+import io.pravega.client.segment.impl.NoSuchSegmentException;
+import io.pravega.client.segment.impl.Segment;
+import io.pravega.client.segment.impl.SegmentInputStreamFactory;
+import io.pravega.client.segment.impl.SegmentMetadataClient;
+import io.pravega.client.segment.impl.SegmentMetadataClientFactory;
+import io.pravega.client.segment.impl.SegmentTruncatedException;
+import io.pravega.client.stream.SegmentReader;
+import io.pravega.client.stream.Serializer;
+import io.pravega.client.stream.TruncatedDataException;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.shared.security.auth.AccessOperation;
+import lombok.extern.slf4j.Slf4j;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeoutException;
+
+@Slf4j
+public class SegmentReaderImpl<T> implements SegmentReader<T> {
+
+    private final Segment segment;
+    private final Serializer<T> deserializer;
+    private final EventSegmentReader input;
+    private final ClientConfig clientConfig;
+    private final SegmentMetadataClient metadataClient;
+
+    public SegmentReaderImpl(SegmentInputStreamFactory factory, Segment segment, Serializer<T> deserializer,
+                             long startOffset, ClientConfig clientConfig, Controller controller,
+                             SegmentMetadataClientFactory segmentMetadataClientFactory) {
+
+        this.segment = segment;
+        this.deserializer = deserializer;
+        this.clientConfig = clientConfig;
+        DelegationTokenProvider delegationTokenProvider = DelegationTokenProviderFactory.create(controller, segment, AccessOperation.READ);
+        metadataClient = segmentMetadataClientFactory.createSegmentMetadataClient(segment,
+                delegationTokenProvider);
+        this.input = factory.createEventReaderForSegment(segment);
+        this.input.setOffset(startOffset);
+    }
+
+    @Override
+    public T read(long timeoutMillis) throws TruncatedDataException, EndOfSegmentException {
+        try {
+            ByteBuffer read = input.read();
+            return read != null ? deserializer.deserialize(read) : null;
+        } catch (NoSuchSegmentException | SegmentTruncatedException e) {
+            handleSegmentTruncated();
+            throw new TruncatedDataException("Segment " + segment + " has been truncated.");
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        return Futures.toVoid(input.fillBuffer());
+    }
+
+    private void handleSegmentTruncated()  {
+        log.info("{} encountered truncation for segment while read{} ", this, segment);
+        try {
+            input.setOffset(Futures.getThrowingExceptionWithTimeout(metadataClient.fetchCurrentSegmentHeadOffset(),
+                    clientConfig.getConnectTimeoutMilliSec()));
```

Comment:
If we encounter NoSuchSegmentException in the caller of handleSegmentTruncated(), what is the expectation here ? what offset are we setting here ?